### PR TITLE
Fix bugs and add tests for sync storage 

### PR DESCRIPTION
**Pull request by barbogast**

### CODE REVIEW

**Comment by byteab** [comment]
Thank you @barbogast I will check that.
maybe I can pick the test part for now.

**Comment by byteab** [comment]
after finalizing the `toThenable`. would you mind to please help on resolving bugs by another PR.
currently tests are also failing for async as well.
if you have time would you please help me to spot why async test is failing.

File: tests/persistSync.test.tsx [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+      onRehydrateStorage: () => (_, e) => {
+        postRehydrationCallbackCallCount++
+        expect(e?.message).toBe('migrate error')
+      },
+    })
+  )
+
+  expect(useStore.getState()).toEqual({ count: 0 })
+  expect(postRehydrationCallbackCallCount).toBe(1)
+})
```
Comment (dai-shi) [inline tests/persistSync.test.tsx:173]: Comparing against https://github.com/pmndrs/zustand/blob/master/tests/persist.test.tsx, this looks good to me.

**Comment by barbogast** [comment]
> would you mind to please help on resolving bugs by another PR.

How about we merge this PR as it is? Then all the relevant code would be in https://github.com/pmndrs/zustand/pull/403 and we would be back to one place in which the conversation happens. Working on `toThenable` should be easier then, because we've got green tests.

**Comment by byteab** [comment]
> > would you mind to please help on resolving bugs by another PR.
> 
> How about we merge this PR as it is? Then all the relevant code would be in [pmndrs#403](https://github.com/pmndrs/zustand/pull/403) and we would be back to one place in which the conversation happens. Working on `toThenable` should be easier then because we've got green tests.

Hmm, I was thinking to first have a final solution on how to implement `toThenable()`, and after that, we start bug fixing. because on the new version of `toThenable()` as provided on gist we handle `try catch`. and some commits on this PR may not be relevant to merge.
what do you think?

**Comment by byteab** [comment]
I will cherry-pick the test part for now.

**Comment by barbogast** [comment]
I guess the tests are the important part, so yeah, I agree. 7052e64 is also relevant, AFAICS, since it fixes the initial loading of the state from localStorage.

**Comment by byteab** [comment]
the tests are now available on `async-await-fix` branch.

**Pull request closed without merging**